Health - body part and organ code separation

Aranclanos opened the pull request:
### Purpose
Cleans up a big chunk of health code, making it simpler and easier to understand without losing any features

Removes RootBodyPartContainer
Removes the usage of dynamic classes
Starts the separation of body parts and organs, with the idea of them being different from eachother. With the health master holding body parts and the body parts holding organs.
Removes organs components from body parts prefabs.
Cleans up a bit gibbing code.
Adds a couple of TODOs to continue the separation and clean-up, as for now organ prefabs still have a body part component.
Fixes #6948

CL:[fix] you can no longer become unkillable by surgically attaching a torso within another torso.

Aranclanos committed:
Removes RootBodyPartContainer
Removes the usage of dynamic classes
Starts the separation of body parts and organs, with the idea of them being different from eachother. With the health master holding body parts and the body parts holding organs.
Removes organs components from body parts prefabs.
Cleans up a bit gibbing code.
Adds a couple of TODOs to continue the separation and clean-up, as for now organ prefabs still have a body part component.

Aranclanos committed:
Fixes organs runtiming on removal.
Fixes some outdated summary comments in body parts.
Fixes tests.
Silences codacy.

NoooneyDude commented on UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:
```diff
@@ -15,52 +16,48 @@ namespace HealthV2
+		[Tooltip("Player sprites for rendering equipment and clothing on this")]
+		public PlayerSprites playerSprites;
+
+		[HideInInspector] public bool IsBleeding = false;
```
```suggestion
		[NonSerialized]
		public bool IsBleeding = false;
```
Guessing we could make this  `[NonSerialized]`? Few others marked `[HideInInspector]` that could probably be treated the same, though no biggie.

MaxIsJoe replied:
that should be a private bool, with a `IsBleeding => isBleeding`

NoooneyDude commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -1260,6 +1147,100 @@ protected virtual void LethalElectrocution(Electrocution electrocution, float sh
+
+				i++;
+				i++;
+				i++;
```
```suggestion
				i += 3;
```
Increment thrice? If that's what you want, perhaps `i += 3` with a quick comment would look less like a mistake.

NoooneyDude commented on UnityProject/Assets/Scripts/HealthV2/Living/Surgery/Procedures/ImplantProcedure.cs:
```diff
@@ -21,14 +21,14 @@ public class ImplantProcedure : SurgeryProcedureBase
 				else
 				{
-					PresentProcedure.ISon.GetComponent<LivingHealthMasterBase>().GetRootBodyPartInZone(interaction.TargetBodyPart).AddBodyPartSlot(interaction.HandSlot);
+					PresentProcedure.ISon.GetComponent<LivingHealthMasterBase>().BodyPartStorage.ServerTryTransferFrom(interaction.HandSlot);
```
`ISon`? Oh, the target body I take it.

I like the `Finnish surgery procedure`, too. 😄

NoooneyDude commented on UnityProject/Assets/Scripts/Items/Medical/Gauze.cs:
```diff
@@ -31,47 +31,41 @@ public override void ServerPerformInteraction(HandApply interaction)
 					{
-						if(limb.IsBleedingExternally)
+						bodyPart.StopExternalBleeding();
+						if(interaction.Performer.Player().GameObject == interaction.TargetObject.Player().GameObject)
```
```suggestion
						if (interaction.Performer == interaction.TargetObject)
```

Aranclanos replied:
the target object seems to be a body part, not the person

NoooneyDude replied:
Not sure I understand your comment. `interaction.Performer` is the performing player's gameobject, and the `Player()` extension returns `ConnectedPlayer`. The `GameObject` member is a reference to the `ConnectedPlayer`'s gameobject, same as previous, so it could just be shortened to `interaction.Performer`.

If you want to call it as outside the scope of the PR as it was already like that, that's fine, but it's a pretty quick change.

Aranclanos replied:
targetobject is a bodypart, and thats a different network object, so player() doesnt get the connection if you use it, I guess that's the reason of why using gauzes on yourself doesnt trigger the proper message

NoooneyDude replied:
Ooh. I getcha. Thanks.

NoooneyDude commented on UnityProject/Assets/Scripts/Items/Medical/Gauze.cs:
```diff
@@ -31,47 +31,41 @@ public override void ServerPerformInteraction(HandApply interaction)
-								$"You stopped {interaction.TargetObject.Player().Script.visibleName}'s bleeding.",
-								$"{interaction.PerformerPlayerScript.visibleName} stopped {interaction.TargetObject.Player().Script.visibleName}'s bleeding.");
-							}
+							Chat.AddActionMsgToChat(interaction.Performer.gameObject,
+							$"You stopped your {interaction.TargetObject.Player().Script.visibleName}'s bleeding.",
+							$"{interaction.PerformerPlayerScript.visibleName} stopped their own bleeding from their {interaction.TargetObject.ExpensiveName()}.");
```
"You stopped your Clarke Hanford's bleeding." 🤔
"Clarke Hanford stopped their own bleeding from their Clarke Hanford." 😆 May as well fix this while it is in the light.
```suggestion
							Chat.AddActionMsgToChat(interaction.Performer,
							$"You stopped the bleeding on your {bodyPart}.",
							$"{interaction.PerformerPlayerScript.visibleName} stopped the bleeding on their {bodyPart}.");
```
Though `bodyPart` probably doesn't print nicely. Does it have an attributes component for `bodyPart.ExpensiveName()`?

Should probably check this before applying suggestion. Same goes for the next.

MaxIsJoe replied:
Like I said on discord, better get the bodyPart's name from the BodyPartDamageDiscreptor script as it's designer defined and constant.

NoooneyDude commented on UnityProject/Assets/Scripts/Items/Medical/Gauze.cs:
```diff
@@ -31,47 +31,41 @@ public override void ServerPerformInteraction(HandApply interaction)
+						}
+						else
+						{
+							Chat.AddActionMsgToChat(interaction.Performer.gameObject,
+							$"You stopped {interaction.TargetObject.Player().Script.visibleName}'s bleeding.",
+							$"{interaction.PerformerPlayerScript.visibleName} stopped {interaction.TargetObject.Player().Script.visibleName}'s bleeding.");
```
```suggestion
							Chat.AddActionMsgToChat(interaction.Performer,
							$"You stopped the bleeding on {interaction.TargetObject.Player().Script.visibleName}'s {bodyPart.ExpensiveName()}.",
							$"{interaction.PerformerPlayerScript.visibleName} stopped the bleeding 
 on {interaction.TargetObject.Player().Script.visibleName}'s {bodyPart.ExpensiveName()}."); // Might read a lil weird for the recipient.
```

NoooneyDude commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -128,13 +121,16 @@ public float MaxHealth
 		/// A list of all body parts of the creature
 		/// </summary>
-		public HashSet<BodyPart> BodyPartList = new HashSet<BodyPart>();
+		public List<BodyPart> BodyPartList = new List<BodyPart>();
```
Just curious, is there something about `HashSet` that means we should avoid it, or were the collection's advantages just not used in this case?

Bod9001 replied:
There were occasions where double body parts ended up getting added so I used HashSet to just prevent that

Aranclanos replied:
this doesn't seem to happen anymore, I'm unaware of the cause of the bug

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/RootBodyPartController.cs:
```diff
@@ -14,49 +15,33 @@ public class RootBodyPartController : NetworkBehaviour
 	private string SerialisedNetIDs = "";
 
-	public List<RootBodyPartContainer> RootBodyParts = new List<RootBodyPartContainer>();
+	public List<uint> ToSerialised = new List<uint>();
```
this stuff is changed in the Unity update

Aranclanos replied:
understandable, however this list was only updated to discard the RootBodyParts

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:
```diff
@@ -273,11 +273,6 @@ public float TotalDamage
 
-		public void DamageInitialisation()
-		{
-			this.AddModifier(DamageModifier);
```
where has this gone?

Aranclanos replied:
was called in awake so moved it there

Bod9001 replied:
can't see in awake in BodyPartDamage.cs

Aranclanos replied:
partial class

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -128,13 +121,16 @@ public float MaxHealth
 		/// </summary>
-		public List<RootBodyPartContainer> RootBodyPartContainers = new List<RootBodyPartContainer>();
+		[HideInInspector]
+		public ItemStorage BodyPartStorage;
```
the player shouldn't have item storage, All they have is dynamic item storage

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -536,7 +512,6 @@ private void SetConsciousState(ConsciousState NewConsciousState)
 					healthStateController.SetConsciousState(NewConsciousState);
 					OnConsciousStateChangeServer.Invoke(oldState, NewConsciousState);
 				}
-				ConsciousState = NewConsciousState;
```
Doesn't this need to be set?

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -554,20 +529,13 @@ private void SetConsciousState(ConsciousState NewConsciousState)
-			foreach (var Container in RootBodyPartContainers)
+			foreach (var bodyPart in BodyPartList)
 			{
-				if (damageSplit)
```
what happened to damage Split?

Aranclanos replied:
since there aren't body parts inside of body parts, the damagesplit dmg value is calculated a few lines above and then applied to all

Bod9001 replied:
there could still be multiple in the same area so for example implanting multiple arms

Aranclanos replied:
solved on last commit

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -801,41 +722,20 @@ public RootBodyPartContainer GetRootBodyPartInZone(BodyPartType bodyPartAim)
-			//TODO: Reimplement
-		}
-
-		public void CheckDismemberBody()
```
Where is this handled?

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -1260,6 +1147,100 @@ protected virtual void LethalElectrocution(Electrocution electrocution, float sh
+			var sprites = implant.GetBodyTypeSprites(playerSprites.ThisCharacter.BodyType);
+			foreach (var Sprite in sprites.Item2)
+			{
+				var newSprite = Spawn.ServerPrefab(implant.SpritePrefab.gameObject, Vector3.zero, playerSprites.BodySprites.transform).GameObject.GetComponent<BodyPartSprites>();
```
note Might have to adjust some sprite rendering orders Since if it'll been thrown on the same object misses separate objects before it might change the rendering order

Aranclanos replied:
thats for another pr focused on player sprites, this particular function was just moved

Bod9001 commented on UnityProject/Assets/Scripts/Items/Implants/Organs/BodyPartUISlots.cs:
```diff
@@ -24,25 +24,6 @@ public class BodyPartUISlots : Organ, IDynamicItemSlotS
 		[SerializeField] [FormerlySerializedAs("Storage")]
 		private List<BodyPartUISlots.StorageCharacteristics> storage;
 
-		public override void RemovedFromBody(LivingHealthMasterBase livingHealthMasterBase)
```
where is this stuff handled now?

Aranclanos replied:
SetHealthMaster() in BodyPart

Bod9001 replied:
why Separate it out?

Aranclanos replied:
more organized that way, plus this is no longer an override since its not longer an organ but a monobehaviour

Bod9001 commented on UnityProject/Assets/Scripts/Items/Implants/Organs/Stomach.cs:
```diff
@@ -47,22 +47,15 @@ public override void ImplantPeriodicUpdate()
-				StartCoroutine(DelayAddFat());
+				var Added = Spawn.ServerPrefab(BodyFatToInstantiate.gameObject).GameObject.GetComponent<BodyFat>();
+				BodyFats.Add(Added);
+				RelatedPart.OrganStorage.ServerTryAdd(Added.gameObject);
```
the problem here is it causes the loop to get an extra element while looping through

MaxIsJoe replied:
I think we need a comment to make sure people are aware of that

Aranclanos replied:
the preriodic update of organs now loops backwards to avoid that

Bod9001 replied:
well what about when you remove element?

Aranclanos replied:
it would runtime, I'll make a reverse for loop

Bod9001 replied:
Wouldn't it still miss an element out?

Aranclanos replied:
no, its the point of the reverse loop -- then again this is handled a bit weirdly already by a disbowel call so it isnt needed

Bod9001 commented on UnityProject/Assets/Scripts/Items/Weapons/Projectiles/Behaviours/ProjectileIncendiary.cs:
```diff
@@ -65,15 +65,9 @@ public bool OnHit(MatrixManager.CustomPhysicsHit hit)
 			if (hit.CollisionHit.GameObject == null) return true;
 
-			//TODO REMOVE AFTER MOBS ARE MOVED TO NEW HEALTH
-			if (hit.CollisionHit.GameObject.TryGetComponent(out LivingHealthMasterBase health))
```
have mobs been moved over to new health?

MaxIsJoe replied:
not yet, but I've been working on making damage work between all mobs again. not just the player.

Aranclanos replied:
this was checking for the new class again a few lines below but yeah I guess I'll add it back with the old class

Bod9001 commented on UnityProject/Assets/Scripts/Player/PlayerSprites.cs:
```diff
@@ -182,68 +174,55 @@ public void SubSetBodyPart(BodyPart Body_Part, string path)
 		if (customisationStorage != null)
 		{
-			bool pass = true;
+			var data = customisationStorage.Data.Replace("@£", "\"");
```
a good point here is that you need customisation for organs to

Aranclanos replied:
there arent many changes on how sprites are handled, mostly its just moving stuff around, like in this case. If the concern is customization for organs (beards, hair, eyes, etc) its working as previously

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -705,12 +673,13 @@ BodyPartType bodyPartAim
+				if (bodyPart.BodyPartType == bodyPartAim)
 				{
 					//Assuming is only going to be one otherwise damage will be duplicated across them
-					bodyPartContainer.TakeDamage(damagedBy, damage, attackType, damageType, armorPenetration);
```
so,  Since this was expecting root body containers, 
the extra logic for having multiple body parts in an area would be handled by them

so, this has to be changed to take account of how many body parts are there, 
so for Bomb , Fire , Rad it splits the damage over all body parts in target area

For everything else it picks a random body part and applies the damage to that

Aranclanos replied:
solved on last commit

Bod9001 commented on UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:
```diff
@@ -554,20 +529,13 @@ private void SetConsciousState(ConsciousState NewConsciousState)
 			}
 
-			foreach (var Container in RootBodyPartContainers)
+			foreach (var bodyPart in BodyPartList)
```
You can't use BodyPartList since that would include organs as well, you need to do damage to the root body parts to propagate the damage down properly through the different armour of the body parts

Aranclanos replied:
BodyPartList doesnt include organs anymore

Bod9001 replied:
how?

Aranclanos replied:
I'm just not adding organs to this list anymore

Aranclanos committed:
Fixes gibbing causing an infinite loop.

Aranclanos committed:
Removes the list ContainBodyParts from BodyParts.

Aranclanos committed:
Fixes organ extraction and implantation.
Fixes brute force disemboweling.

Aranclanos committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into bodypartcontainerremoval

Aranclanos committed:
Reduces some bloat in LivingHealthMasterBase regarding apply damage functions.
Damage targeted to one body part type will now spread across all of the same type instead of picking the first one.

Aranclanos committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into bodypartcontainerremoval

Aranclanos committed:
Merge branch 'develop' of https://github.com/unitystation/unitystation into bodypartcontainerremoval

Aranclanos committed:
Merge conflicts

Aranclanos committed:
Merge branch 'develop' into bodypartcontainerremoval

Bod9001 reviewed (approved):
apart from the few questions it looks good

corp-0 reviewed (approved):
after this approval, only one unresolved conversation left to merge.
